[SPIR-V] Add pass to merge convergence region exit targets

Keenuts opened the pull request:
The structurizer required regions to be SESE: single entry, single exit.
This new pass transforms multiple-exit regions into single-exit regions.

```
      +---+
      | A |
      +---+
      /   \
   +---+ +---+
   | B | | C |  A, B & C belongs to the same convergence region.
   +---+ +---+
     |     |
   +---+ +---+
   | D | | E |  C & D belongs to the parent convergence region.
   +---+ +---+  This means B & C are the exit blocks of the region.
      \   /     And D & E the targets of those exits.
       \ /
        |
      +---+
      | F |
      +---+
```

This pass would assign one value per exit target:
B = 0
C = 1

Then, create one variable per exit block (B, C), and assign it to the correct value: in B, the variable will have the value 0, and in C, the value 1.

Then, we'd create a new block H, with a PHI node to gather those 2 variables, and a switch, to route to the correct target.

Finally, the branches in B and C are updated to exit to this new block.

```
      +---+
      | A |
      +---+
      /   \
   +---+ +---+
   | B | | C |
   +---+ +---+
      \   /
      +---+
      | H |
      +---+
      /   \
   +---+ +---+
   | D | | E |
   +---+ +---+
      \   /
       \ /
        |
      +---+
      | F |
      +---+
```

Note: the variable is set depending on the condition used to branch. If B's terminator was conditional, the variable would be set using a SELECT.
All internal edges of a region are left intact, only exiting edges are updated.

Keenuts committed:
[SPIR-V] Add pass to merge convergence region exit targets

The structurizer required regions to be SESE: single entry, single
exit.
This new pass transforms multiple-exit regions into single-exit regions.

```
      +---+
      | A |
      +---+
      /   \
   +---+ +---+
   | B | | C |  A, B & C belongs to the same convergence region.
   +---+ +---+
     |     |
   +---+ +---+
   | D | | E |  C & D belongs to the parent convergence region.
   +---+ +---+  This means B & C are the exit blocks of the region.
      \   /     And D & E the targets of those exits.
       \ /
        |
      +---+
      | F |
      +---+
```

This pass would assign one value per exit target:
B = 0
C = 1

Then, create one variable per exit block (B, C), and assign it
to the correct value: in B, the variable will have the value 0,
and in C, the value 1.

Then, we'd create a new block H, with a PHI node to gather those
2 variables, and a switch, to route to the correct target.

Finally, the branches in B and C are updated to exit to this new block.

```
      +---+
      | A |
      +---+
      /   \
   +---+ +---+
   | B | | C |
   +---+ +---+
      \   /
      +---+
      | H |
      +---+
      /   \
   +---+ +---+
   | D | | E |
   +---+ +---+
      \   /
       \ /
        |
      +---+
      | F |
      +---+
```

Note: the variable is set depending on the condition used to branch.
If B's terminator was conditional, the variable would be set using a
SELECT.
All internal edges of a region are left intact, only exiting edges are
updated.

Signed-off-by: Nathan Gauër <brioche@google.com>

iliya-diyachkov commented on llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp:
```diff
@@ -1074,6 +1084,10 @@ void SPIRVEmitIntrinsics::processGlobalValue(GlobalVariable &GV,
 void SPIRVEmitIntrinsics::insertAssignPtrTypeIntrs(Instruction *I,
                                                    IRBuilder<> &B) {
+  // Don't assign types to LLVM tokens.
+  if (isConvergenceIntrinsic(I))
```
Maybe move this check to SPIRVEmitIntrinsics::runOnFunction(), where insertAssignPtrTypeIntrs() is called.

Keenuts replied:
Yes, that's better, changed, thanks!

iliya-diyachkov commented on llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp:
```diff
@@ -1092,6 +1106,10 @@ void SPIRVEmitIntrinsics::insertAssignPtrTypeIntrs(Instruction *I,
 void SPIRVEmitIntrinsics::insertAssignTypeIntrs(Instruction *I,
                                                 IRBuilder<> &B) {
+  // Don't assign types to LLVM tokens.
+  if (isConvergenceIntrinsic(I))
```
The same suggestion to move to SPIRVEmitIntrinsics::runOnFunction().

iliya-diyachkov commented on llvm/lib/Target/SPIRV/SPIRVMergeRegionExitTargets.cpp:
```diff
@@ -0,0 +1,290 @@
+
+    if (auto *SI = dyn_cast<SwitchInst>(T)) {
+      output.insert(SI->getDefaultDest());
+      for (auto &Case : SI->cases()) {
+        output.insert(Case.getCaseSuccessor());
+      }
```
Maybe don't need braces here.

iliya-diyachkov commented on llvm/lib/Target/SPIRV/SPIRVMergeRegionExitTargets.cpp:
```diff
@@ -0,0 +1,290 @@
+      BasicBlock *BB,
+      const std::unordered_map<BasicBlock *, ConstantInt *> &TargetToValue) {
+    auto *T = BB->getTerminator();
+    if (auto *RI = dyn_cast<ReturnInst>(T)) {
+      return nullptr;
+    }
```
Maybe don't need braces here.

Keenuts committed:
pr-feedback

Signed-off-by: Nathan Gauër <brioche@google.com>

Keenuts commented:
Thanks, feedback applied!

Keenuts committed:
remove F.dump() call

sudonatalie commented on llvm/lib/Target/SPIRV/SPIRVMergeRegionExitTargets.cpp:
```diff
@@ -0,0 +1,287 @@
+      modified = true;
+    }
+
+#if !NDEBUG
```
Should we gate this behind `EXPENSIVE_CHECKS` instead/in addition? (I can't remember if it's default true for debug builds)

Keenuts replied:
Seems like some passes enable it on both NDEBUG and EXPENSIVE_CHECKS, and some only EXPENSIVE_CHECKS. Since this is not that costly to do, I'll enable it in both cases (at least until this structurizer is more stable/robust).

sudonatalie commented on llvm/lib/Target/SPIRV/SPIRVMergeRegionExitTargets.cpp:
```diff
@@ -0,0 +1,287 @@
+
+namespace llvm {
+void initializeSPIRVMergeRegionExitTargetsPass(PassRegistry &);
+} // namespace llvm
```
nit: merge same side-by-side namespace blocks?

sudonatalie commented on llvm/lib/Target/SPIRV/SPIRVTargetMachine.cpp:
```diff
@@ -164,6 +164,7 @@ void SPIRVPassConfig::addIRPasses() {
     //  - all loop exits are dominated by the loop pre-header.
     //  - loops have a single back-edge.
     addPass(createLoopSimplifyPass());
+    addPass(createSPIRVMergeRegionExitTargetsPass());
```
Add another pass summary comment like the one above?

Keenuts replied:
done thanks!

Keenuts committed:
pr feedback

Keenuts committed:
pr feedback, merge namespaces

michalpaszkowski reviewed (approved):
@Keenuts Thank you for the patch and detailed explanation! LGTM! 
A high level note for either this/future patch: It may be worth to update the [SPIRVUsage document](https://llvm.org/docs/SPIRVUsage.html) with how target intrinsics are affected in the Vulkan environment. We can discuss if there is any sensible way to split this document to discuss differences between SPIR-V "flavors".

Keenuts commented:
> @Keenuts Thank you for the patch and detailed explanation! LGTM! A high level note for either this/future patch: It may be worth to update the [SPIRVUsage document](https://llvm.org/docs/SPIRVUsage.html) with how target intrinsics are affected in the Vulkan environment. We can discuss if there is any sensible way to split this document to discuss differences between SPIR-V "flavors".

Oh right, yes, could be valuable to add this, and in general, explain how the LLVM IR is structurized for the graphical SPIR-V.

MaskRay commented:
`auto *RI = dyn_cast` when `RI` is unused causes -Wunused-but-set-variable warnings with a relative new Clang (>= 2021-04). I fixed them in a088c61d5c409ec0b8994340866d5864ba913516.

You probably want to compile and test a modern Clang to catch such issues.

Keenuts commented:
> `auto *RI = dyn_cast` when `RI` is unused causes -Wunused-but-set-variable warnings with a relative new Clang (>= 2021-04). I fixed them in [a088c61](https://github.com/llvm/llvm-project/commit/a088c61d5c409ec0b8994340866d5864ba913516).
> 
> You probably want to compile and test a modern Clang to catch such issues.

Oh, thanks for fixing!
Weirdly my system's clang is 16.0.6, so seems like it's from 2023.
Added '-DCMAKE_CXX_FLAGS=-Wunused-but-set-variable" but this warning is still not showing on main (before your PR of course). With what options do you compile?

Mine are:
```
cmake -Hllvm -GNinja -Bbuild -DCMAKE_BUILD_TYPE=Debug -DLLVM_ENABLE_PROJECTS='clang' -DLLVM_TARGETS_TO_BUILD='X86' -DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD="DirectX;SPIRV" -DLLVM_OPTIMIZED_TABLEGEN=1 -DLLVM_ENABLE_LLD=1 -DLLVM_USE_SPLIT_DWARF=1 -DCMAKE_CXX_COMPILER=clang++-16 -DCMAKE_C_COMPILER=clang-16 -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -DCMAKE_INSTALL_PREFIX=build/install -DLLVM_INCLUDE_SPIRV_TOOLS_TESTS=1
```

VyacheslavLevytskyy commented on llvm/test/CodeGen/SPIRV/structurizer/merge-exit-multiple-break.ll:
```diff
@@ -0,0 +1,103 @@
+
+; CHECK:   %[[#new_end]] = OpLabel
+; CHECK:    %[[#route:]] = OpPhi %[[#int_ty]] %[[#int_2]] %[[#while_cond]] %[[#int_0]] %[[#while_body]] %[[#int_1]] %[[#if_end]]
+; CHECK:                   OpSwitch %[[#route]] %[[#while_end_loopexit]] 1 %[[#if_then2]] 0 %[[#if_then]]
```
@Keenuts It looks like this line turns the test case into a flaky one. I run it multiple times and get
* either `; CHECK:                   OpSwitch %[[#route]] %[[#if_then2]] 2 %[[#while_end_loopexit]] 0 %[[#if_then]]`
* or current one (`OpSwitch %[[#route]] %[[#while_end_loopexit]] 1 %[[#if_then2]] 0 %[[#if_then]]`)
randomly.

VyacheslavLevytskyy replied:
Sometimes it requires tens of runs, but eventually you get the other option instead of the current merged line.

Keenuts replied:
Ahh, I had those issues when writing the PR, so made sure not to use `SmallPtrSet` anywhere around those, but might have missed a case.
I would like to keep the constant lookup time for most cases, so maybe I should add a step to reorder the switch operands based on something more deterministic, like basic-block order. Looking into that today.

ssahasra reviewed (commented):
There is a lot of overlap between GPU backends like SPIRV and AMDGPU, where we are tackling similar problems with the CFG. It will be good to reuse and improve code wherever possible. After a cursory look, I think `FixIrreducible.cpp` and `UnifyLoopExits.cpp` can both potentially benefit from the "exit variable" created here. Those passes currently track a whole bunch of boolean variables, one for each exit block.
